Grant/contracts/BuilderGrant.sol

```diff
@@ -0,0 +1,334 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity ^0.8.26;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+
+interface IMCV2_Bond {
+    function mint(
+        address token,
+        uint256 tokensToMint,
+        uint256 maxReserveAmount,
+        address receiver
+    ) external returns (uint256);
+}
+
+/**
+ * @title BuilderGrant
+ * @author sebayaki (@if)
+ * @notice This contract is used to distribute grants to Builders (Tip Receivers)
+ * @dev
+ * - The top 3 builders will claim bounties of X, Y, Z HUNT each, in the form of Mini Building NFTs (each backed by 100 HUNT tokens).
+ * - The top 3 builders have 3 options to choose from:
+ *   - Option 1: Claim 100% for themselves.
+ *   - Option 2: Claim 50% for themselves and donate 50% to those ranked 4th and below.
+ *   - Option 3: Donate 100% to those ranked 4th and below.
+ * - All data is stored on the contract, so `setSeasonData` requires the owners to add at least up to the (X / 100 + 3)rd rank to cover all donations.
+ * - Donation receivers will be able to claim their rewards once the top 3 finish their decisions or after 1 week has passed.
+ */
+
+contract BuilderGrant is Ownable {
+    error InvalidSeasonId();
+    error InvalidRankersParams();
+    error InvalidGrantAmount();
+    error NothingToWithdraw();
+    error ClaimDeadlineReached();
+    error DonationClaimDeadlineReached();
+    error DonationNotClaimableYet();
+    error SeasonDataIsNotUpdateable();
+    error NotEnoughGrantBalance();
+    error TokenTransferFailed();
+    error InvalidRankingParam();
+    error PermissionDenied();
+    error NotARanker();
+    error AlreadyClaimed();
+    error DonationAlreadyClaimed();
+    error InvalidClaimType();
+    error MintBuildingsFailed();
+    error NoDonationToClaim();
+
+    uint256 public constant HUNT_PER_MINI_BUILDING = 100 ether; // 100 HUNT per Mini Building NFT minting
+
+    uint256 public constant TOP3_CLAIM_DEADLINE = 1 weeks;
+    uint256 public constant DONATION_CLAIM_DEADLINE = 4 weeks; // 2 - 4 weeks after TOP3_CLAIM_DEADLINE
+
+    IERC20 public immutable HUNT;
+    IMCV2_Bond public immutable BOND;
+    address public immutable MINI_BUILDING_ADDRESS;
+
+    struct Ranker {
+        uint48 fid; // INFO: Farcaster ID
+        address wallet;
+        uint16 claimedAmount;
+        bool[3] donationReceived;
+    }
+
+    struct Grant {
+        uint8 claimedType; // 0: not claimed yet, 1: 100% self, 2: 50% donation, 3: 100% donation
+        uint16 amount; // mini building count
```

Comment:
changing from 0 -> n is more gas than 1 -> n, the types imo should start at 1, so be 1,2,3,4

see https://chatgpt.com/share/422a6932-d69e-4381-9606-8cb1d40f0026

Reply:
Then the owner should set all `claimType` to 1 as default value in `setSeasonData`. This makes the gas a little more expensive for the owners and a little cheaper when the top 3 users claim. However, it will increase the total gas spending if we consider both the owner and the winners.